File: packages/engine/src/core/file-mentions.ts
/**
 * File mentions (memory-lead 2.1, D6): which files a decision's own text
 * names. A mention is the derived half of a decision's scope, where a `path:`
 * guard is the declared half.
 *
 * A mention states only that a decision NAMES a file, never that it governs
 * it (record-index D2). So the extractor may over-collect: a token that is
 * not a file (`Node.js`, `api.sofar.sh`) matches only a path that ends in it,
 * and no real read does. What it must not do is match broadly. Directory
 * tokens are therefore not mentions: on this repo they alone spread 92
 * decisions over 375 files, where file tokens name 74, with a median of one
 * decision per file. Directory scope is what a `path:` guard declares.
 *
 * Pure and deterministic, like rule-fidelity's classifier, whose file-name
 * class this reuses for the last segment.
 */

/** Delimiters a path never contains: whitespace, backticks, quotes, brackets, commas, semicolons. */
const SPLIT_RE = /[\s`"“”‘’'(),;<>[\]{}|]+/
const TRAILING_PUNCT_RE = /[.:!?]+$/
/** `src/a.ts:42`, `src/a.ts:42:7` and `src/a.ts#L10-L12` name the file. */
const LOCATION_RE = /(?::\d+(?::\d+)?|#L\d+(?:-L?\d+)?)$/
/**
 * rule-fidelity's file-name class (`categories.ts`), or a dotfile (`.mcp.json`,
 * `.gitignore`). A bare name needs two characters before its extension, which
 * keeps `e.g` out; behind a `/` one will do (`src/a.ts`).
 */
const FILE_SEGMENT_RE = /^(?:[\w-]{2,}(?:\.[\w-]+)*\.[A-Za-z][A-Za-z0-9]{0,4}|\.[\w-][\w.-]*)$/
const PATHED_FILE_SEGMENT_RE = /^(?:[\w-]+(?:\.[\w-]+)*\.[A-Za-z][A-Za-z0-9]{0,4}|\.[\w-][\w.-]*)$/
const PATH_SEGMENT_RE = /^[\w.@+-]+$/

/** The file tokens of `text`, in order, deduplicated. */
export function fileMentions(text: string): string[] {
  const found: string[] = []
  const seen = new Set<string>()
  for (const raw of text.split(SPLIT_RE)) {
    const token = fileToken(raw)
    if (token === null || seen.has(token)) continue
    seen.add(token)
    found.push(token)
  }
  return found
}

function fileToken(raw: string): string | null {
  let token = raw.replace(TRAILING_PUNCT_RE, '').replace(LOCATION_RE, '').replace(TRAILING_PUNCT_RE, '')
  const pathed = token.includes('/')
  if (token.startsWith('./')) token = token.slice(2)
  if (token.length === 0 || token.includes('://') || /[*?$]/.test(token) || token.startsWith('~')) return null
  const segments = (token.startsWith('/') ? token.slice(1) : token).split('/')
  const last = segments[segments.length - 1] ?? ''
  if (!(pathed ? PATHED_FILE_SEGMENT_RE : FILE_SEGMENT_RE).test(last)) return null
  for (const segment of segments) {
    if (segment.length === 0 || !PATH_SEGMENT_RE.test(segment)) return null
  }
  return token
}

/**
 * How many path segments of `path` the token names, or 0 when it names none
 * of it. A token names a path when it IS the path or its tail at a `/`
 * boundary, so `core/fold.ts` names `/repo/src/core/fold.ts` (2) and
 * `fold.ts` names it too (1), but `old.ts` does not. The count ranks a
 * specific mention over a bare file name.
 */
export function mentionDepth(token: string, path: string): number {
  if (path !== token && !path.endsWith(token.startsWith('/') ? token : `/${token}`)) return 0
  return token.split('/').filter((s) => s.length > 0).length
}

File: crates/sofar-core/src/file_mentions.rs
//! File mentions (`core/file-mentions.ts`, memory-lead 2.1, D6): which files a
//! decision's own text names — the derived half of a decision's scope, where
//! a `path:` guard is the declared half. A mention says a decision NAMES a
//! file, never that it governs it (record-index D2).
//!
//! The TypeScript regexes are ASCII classes in non-unicode mode (`\w` is
//! `[A-Za-z0-9_]`, `\d` is `[0-9]`), so they port as byte tests; `\s` is the
//! JavaScript whitespace class (HOTPATH §Text-semantics pins).

use crate::text::is_js_whitespace;

/// `SPLIT_RE`: whitespace, backticks, quotes, brackets, commas, semicolons.
fn is_split(c: char) -> bool {
    is_js_whitespace(c)
        || matches!(
            c,
            '`' | '"'
                | '\u{201C}'
                | '\u{201D}'
                | '\u{2018}'
                | '\u{2019}'
                | '\''
                | '('
                | ')'
                | ','
                | ';'
                | '<'
                | '>'
                | '['
                | ']'
                | '{'
                | '}'
                | '|'
        )
}

fn is_word(c: char) -> bool {
    c.is_ascii_alphanumeric() || c == '_'
}

/// `[\w-]+` (at least `min` long).
fn word_run(s: &str, min: usize) -> bool {
    s.len() >= min && s.chars().all(|c| is_word(c) || c == '-')
}

/// `[A-Za-z][A-Za-z0-9]{0,4}`.
fn extension(s: &str) -> bool {
    let mut chars = s.chars();
    chars.next().is_some_and(|c| c.is_ascii_alphabetic())
        && s.len() <= 5
        && chars.all(|c| c.is_ascii_alphanumeric())
}

/// `FILE_SEGMENT_RE` (`first_min` 2) and `PATHED_FILE_SEGMENT_RE` (1):
/// `[\w-]{min,}(?:\.[\w-]+)*\.ext`, or a dotfile `\.[\w-][\w.-]*`.
fn file_segment(s: &str, first_min: usize) -> bool {
    if let Some(rest) = s.strip_prefix('.') {
        let mut chars = rest.chars();
        return chars.next().is_some_and(|c| is_word(c) || c == '-')
            && chars.all(|c| is_word(c) || c == '.' || c == '-');
    }
    let parts: Vec<&str> = s.split('.').collect();
    let Some((last, init)) = parts.split_last() else {
        return false;
    };
    !init.is_empty()
        && word_run(init[0], first_min)
        && init[1..].iter().all(|p| word_run(p, 1))
        && extension(last)
}

/// `PATH_SEGMENT_RE`: `^[\w.@+-]+$`.
fn path_segment(s: &str) -> bool {
    !s.is_empty()
        && s.chars()
            .all(|c| is_word(c) || matches!(c, '.' | '@' | '+' | '-'))
}

/// `TRAILING_PUNCT_RE`: `/[.:!?]+$/` replaced by nothing.
fn strip_trailing_punct(s: &str) -> &str {
    s.trim_end_matches(['.', ':', '!', '?'])
}

/// `^\d+`: the length of a leading ASCII digit run.
fn digits(s: &str) -> usize {
    s.bytes().take_while(u8::is_ascii_digit).count()
}

/// Whether `s` is wholly `:\d+(?::\d+)?` or `#L\d+(?:-L?\d+)?`.
fn is_location(s: &str) -> bool {
    if let Some(rest) = s.strip_prefix(':') {
        let n = digits(rest);
        if n == 0 {
            return false;
        }
        let rest = &rest[n..];
        return rest.is_empty()
            || rest
                .strip_prefix(':')
                .is_some_and(|r| !r.is_empty() && digits(r) == r.len());
    }
    let Some(rest) = s.strip_prefix("#L") else {
        return false;
    };
    let n = digits(rest);
    if n == 0 {
        return false;
    }
    let rest = &rest[n..];
    if rest.is_empty() {
        return true;
    }
    let Some(rest) = rest.strip_prefix('-') else {
        return false;
    };
    let rest = rest.strip_prefix('L').unwrap_or(rest);
    !rest.is_empty() && digits(rest) == rest.len()
}

/// `LOCATION_RE` replaced by nothing: the leftmost start whose suffix is
/// wholly a location, as a JavaScript search anchored at `$` finds it.
fn strip_location(s: &str) -> &str {
    for (i, c) in s.char_indices() {
        if (c == ':' || c == '#') && is_location(&s[i..]) {
            return &s[..i];
        }
    }
    s
}

/// `fileToken`.
fn file_token(raw: &str) -> Option<&str> {
    let mut token = strip_trailing_punct(strip_location(strip_trailing_punct(raw)));
    let pathed = token.contains('/');
    if let Some(rest) = token.strip_prefix("./") {
        token = rest;
    }
    if token.is_empty()
        || token.contains("://")
        || token.contains(['*', '?', '$'])
        || token.starts_with('~')
    {
        return None;
    }
    let segments: Vec<&str> = token
        .strip_prefix('/')
        .unwrap_or(token)
        .split('/')
        .collect();
    let last = segments.last().copied().unwrap_or("");
    if !file_segment(last, if pathed { 1 } else { 2 }) {
        return None;
    }
    segments.iter().all(|s| path_segment(s)).then_some(token)
}

/// `fileMentions`: the file tokens of `text`, in order, deduplicated.
#[must_use]
pub fn file_mentions(text: &str) -> Vec<String> {
    let mut found: Vec<String> = Vec::new();
    for raw in text.split(is_split) {
        if let Some(token) = file_token(raw)
            && !found.iter().any(|f| f == token)
        {
            found.push(token.to_owned());
        }
    }
    found
}

/// `mentionDepth`: how many path segments of `path` the token names — it IS
/// the path or its tail at a `/` boundary — or 0 when it names none of it.
#[must_use]
pub fn mention_depth(token: &str, path: &str) -> usize {
    let tail = if token.starts_with('/') {
        token.to_owned()
    } else {
        format!("/{token}")
    };
    if path != token && !path.ends_with(&tail) {
        return 0;
    }
    token.split('/').filter(|s| !s.is_empty()).count()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn tokens_as_the_typescript_extractor_finds_them() {
        assert_eq!(
            file_mentions(
                "edit `src/core/fold.ts:42` and README.md, not e.g. or Node.js; see ./a.ts#L10-L12 and .mcp.json"
            ),
            [
                "src/core/fold.ts",
                "README.md",
                "Node.js",
                "a.ts",
                ".mcp.json"
            ]
        );
        assert_eq!(
            file_mentions("a/b/ src/*.ts ~/x.ts http://x.io/a.ts x.toolongext"),
            Vec::<String>::new()
        );
        assert_eq!(file_mentions("a:1:2:3.ts"), Vec::<String>::new());
        assert_eq!(file_mentions("docs/SPEC.md."), ["docs/SPEC.md"]);
    }

    #[test]
    fn depth_counts_the_named_tail() {
        assert_eq!(mention_depth("core/fold.ts", "/repo/src/core/fold.ts"), 2);
        assert_eq!(mention_depth("fold.ts", "/repo/src/core/fold.ts"), 1);
        assert_eq!(mention_depth("old.ts", "/repo/src/core/fold.ts"), 0);
    }
}
